clientUtils/url.ts

```diff
@@ -1,47 +1,67 @@
-import { isEmpty } from "./Util"
+import { isEmpty, omitUndefinedValues } from "./Util"
 
 export interface QueryParams {
     [key: string]: string | undefined
 }
 
+/**
+ * `_original` contains the original, URI-encoded query param as it is in the URL.
+ *   It should only be used if necessary, e.g. for legacy reasons (we need to
+ *   distinguish between `+` and `%20` for legacy URLs, for example).
+ * `decoded` contains the URL-decoded version of the query params instead.
+ */
+export interface EncodedDecodedQueryParams {
+    _original: QueryParams
+    decoded: QueryParams
+}
+
 // Deprecated. Use getWindowQueryParams() to get the params from the global URL,
 // or strToQueryParams(str) to parse an arbtirary query string.
-export const getQueryParams = (queryStr?: string): QueryParams =>
+export const getQueryParams = (queryStr?: string): EncodedDecodedQueryParams =>
     strToQueryParams(queryStr || getWindowQueryStr())
 
-export const getWindowQueryParams = (): QueryParams =>
+export const getWindowQueryParams = (): EncodedDecodedQueryParams =>
     strToQueryParams(getWindowQueryStr())
 
-export const strToQueryParams = (queryStr = ""): QueryParams => {
+/**
+ * Converts a query string into an object of key-value pairs.
+ * Handles URI-decoding of the values.
+ */
+export const strToQueryParams = (queryStr = ""): EncodedDecodedQueryParams => {
     if (queryStr[0] === "?") queryStr = queryStr.substring(1)
 
     const querySplit = queryStr.split("&").filter((s) => !isEmpty(s))
-    const params: QueryParams = {}
+    const params: EncodedDecodedQueryParams = { _original: {}, decoded: {} }
 
     for (const param of querySplit) {
-        const pair = param.split("=")
-        params[pair[0]] = pair[1]
+        const [key, value] = param.split("=", 2)
+        const decodedKey = decodeURIComponent(key.replace(/\+/g, "%20"))
+        const decoded =
+            value !== undefined
+                ? decodeURIComponent(value.replace(/\+/g, "%20"))
+                : undefined
+
+        params._original[decodedKey] = value
```

Comment:
Just curious how you decided to use `decodedKey` instead of `key` as a key?

Reply:
Ah I see it makes sense to standardize the the keys since they are not causing issues, so you don't have yet another function to match unencoded with the encoded (e.g. have to remember the `+` and `%20` ambiguity, not just a simple `decodeURIComponent`)

Reply:
Ah, right, interesting question!
The reason for this is that it ensures that `queryParamsToUrl(urlToQueryParams(url).decoded) == url` (since `URLSearchParams` encodes keys and values).

If it needs or even should be done for `_original` is another question really, but let's keep it like that for consistency.